File: java/yaskawa/ext/Robot.java
package yaskawa.ext;

import java.util.*;
import java.util.function.*;

import org.apache.thrift.TException;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.protocol.TMultiplexedProtocol;

import yaskawa.ext.api.*;


public class Robot 
{
    Robot(Controller c, TProtocol protocol, int index) throws TTransportException
    {
        this.c = c;
        this.index = index;
        client = new yaskawa.ext.api.Robot.Client(protocol);
    }

    public String model() throws TException
    {
        synchronized(c.extension) {
            return client.model(index);
        }
    }

    public int dof() throws TException
    {
        synchronized(c.extension) {
            return client.dof(index);
        }
    }

    public Position jointPosition(OrientationUnit unit) throws TException
    {
        synchronized(c.extension) {
            return client.jointPosition(index, unit);
        }
    }

    public Position toolTipPosition(CoordinateFrame frame, int tool) throws TException
    {
        synchronized(c.extension) {
            return client.toolTipPosition(index, frame, tool);
        }
    }


    public boolean forceLimitingAvailable() throws TException
    {
        synchronized(c.extension) {
            return client.forceLimitingAvailable(index);
        }
    }

    public boolean forceLimitingActive() throws TException
    {
        synchronized(c.extension) {
            return client.forceLimitingActive(index);
        }
    }

    public boolean forceLimitingStopped() throws TException
    {
        synchronized(c.extension) {
            return client.forceLimitingStopped(index);
        }
    }

    public boolean switchBoxAvailable() throws TException
    {
        synchronized(c.extension) {
            return client.switchBoxAvailable(index);
        }
    }

    public int activeTool() throws TException
    {
        synchronized(c.extension) {
            return client.activeTool(index);
        }
    }

    public void setActiveTool(int tool) throws TException
    {
        synchronized(c.extension) {
            client.setActiveTool(index, tool);
        }
    }

    public Position workHomePosition() throws TException
    {
        synchronized(c.extension) {
            return client.workHomePosition(index);
        }
    }

    public void setWorkHomePosition(Position pos) throws TException
    {
        synchronized(c.extension) {
            client.setWorkHomePosition(index, pos);
        }
    }

    public Position secondHomePosition() throws TException
    {
        synchronized(c.extension) {
            return client.secondHomePosition(index);
        }
    }

    public void setSecondHomePosition(Position pos) throws TException
    {
        synchronized(c.extension) {
            client.setSecondHomePosition(index, pos);
        }
    }


    protected Controller c;
    protected yaskawa.ext.api.Robot.Client client;
    protected int index;
}

File: csharp/Yaskawa/Ext/Robot.cs
using System;
using Thrift.Protocol;
using Yaskawa.Ext.API;

namespace Yaskawa.Ext
{
    public class Robot 
    {
        public Robot(Controller c, Extension ext, TProtocol protocol, int index)
        {
            this.c = c;
            this.index = index;
            extension = ext;


            lock (extension.SyncRoot)
                client = new API.Robot.Client(protocol);
        }
        
        public String model()
        {
            lock (extension.SyncRoot)
                return client.model(index).Result;
        }

        public int dof()
        {
            lock (extension.SyncRoot)
                return client.dof(index).Result;
        }

        public Position jointPosition(OrientationUnit unit)
        {
            lock (extension.SyncRoot)
                return client.jointPosition(index, unit).Result;
        }

        public Position toolTipPosition(CoordinateFrame frame, int tool)
        {
            lock (extension.SyncRoot)
                return client.toolTipPosition(index, frame, tool).Result;
        }


        public bool forceLimitingAvailable()
        {
            lock (extension.SyncRoot)
                return client.forceLimitingAvailable(index).Result;
        }

        public bool forceLimitingActive()
        {
            lock (extension.SyncRoot)
                return client.forceLimitingActive(index).Result;
        }

        public bool forceLimitingStopped()
        {
            lock (extension.SyncRoot)
                return client.forceLimitingStopped(index).Result;
        }

        public bool switchBoxAvailable()
        {
            lock (extension.SyncRoot)
                return client.switchBoxAvailable(index).Result;
        }

        public int activeTool()
        {
            lock (extension.SyncRoot)
                return client.activeTool(index).Result;
        }

        public void setActiveTool(int tool)
        {
            lock (extension.SyncRoot)
                client.setActiveTool(index, tool).Wait();
        }

        public Position workHomePosition()
        {
            lock (extension.SyncRoot)
                return client.workHomePosition(index).Result;
        }

        public void setWorkHomePosition(Position pos)
        {
            lock (extension.SyncRoot)
                client.setWorkHomePosition(index, pos).Wait();
        }

        public Position secondHomePosition()
        {
            lock (extension.SyncRoot)
                return client.secondHomePosition(index).Result;
        }

        public void setSecondHomePosition(Position pos)
        {
            lock (extension.SyncRoot)
                client.setSecondHomePosition(index, pos).Wait();
        }

        public double maximumLinearSpeed()
        {
            lock (extension.SyncRoot)
                return client.maximumLinearSpeed(index).Result;
        }


        protected Controller c;
        protected API.Robot.Client client;
        protected int index;

        protected object SyncRoot
        {
            get;
            private set;
        }
        Extension extension;
    }
}